[4782] Add <- and -> (forward and back) buttons to the URL/internal browser in Preview Toolbar

aachavarria opened the pull request:
https://github.com/craftercms/craftercms/issues/4782

aachavarria committed:
adding Back and Forward buttons to PreviewAddressBar.tsx;
creating actions for goToNextPage and goToLastPage;
Creating Hooks to get the back/forward stacks;

rart commented on ui/app/src/components/BackButtonStack/BackButtonStack.tsx:
```diff
@@ -0,0 +1,40 @@
+import Tooltip from '@material-ui/core/Tooltip';
+import { goToLastPage } from '../../state/actions/preview';
+
+export default function BackButtonStack() {
```
extend iconbutton props and spread them on to the button

rart commented on ui/app/src/components/BackButtonStack/BackButtonStack.tsx:
```diff
@@ -0,0 +1,40 @@
+      <Tooltip title={<FormattedMessage id="backButtonStack.back" defaultMessage="Back" />}>
+        <ArrowBackIosRoundedIcon />
+      </Tooltip>
+    </IconButton>
```
```suggestion
    <Tooltip title={<FormattedMessage id="backButtonStack.back" defaultMessage="Back" />}>
      <span>
        <IconButton disabled={stack.length <= 1} onClick={onClick}>
          <ArrowBackIosRoundedIcon />
        </IconButton>
      </span>
    </Tooltip>
```

rart commented on ui/app/src/components/BackButtonStack/index.ts:
```diff
@@ -0,0 +1,19 @@
+/*
```
Rename all this stuff to PreviewBackButton

rart commented on ui/app/src/components/ForwardButtonStack/ForwardButtonStack.tsx:
```diff
@@ -0,0 +1,40 @@
+/*
```
Rename all this stuff to PreviewForwardButton

rart commented on ui/app/src/components/ForwardButtonStack/ForwardButtonStack.tsx:
```diff
@@ -0,0 +1,40 @@
+      <Tooltip title={<FormattedMessage id="forwardButtonStack.forward" defaultMessage="Forward" />}>
+        <ArrowForwardIosRoundedIcon />
+      </Tooltip>
+    </IconButton>
```
Same deal as the back button

rart commented on ui/app/src/models/GlobalState.ts:
```diff
@@ -153,6 +153,8 @@ export interface GlobalState {
     pageBuilderPanelWidth: number;
     pageBuilderPanelStack: WidgetDescriptor[];
+    backStack: string[];
+    forwardStack: string[];
```
```suggestion
    historyBackStack: string[];
    historyForwardStack: string[];
```

rart commented on ui/app/src/utils/hooks/useBackStack.ts:
```diff
@@ -0,0 +1,22 @@
+import GlobalState from '../../models/GlobalState';
+import { useSelector } from 'react-redux';
+
+export function useBackStack(): GlobalState['preview']['backStack'] {
```
useHistoryBackStack

rart commented on ui/app/src/utils/hooks/useForwardStack.ts:
```diff
@@ -0,0 +1,22 @@
+import GlobalState from '../../models/GlobalState';
+import { useSelector } from 'react-redux';
+
+export function useForwardStack(): GlobalState['preview']['forwardStack'] {
```
useHistoryForwardStack

aachavarria committed:
Extending Back/Forward Buttons from IconButtonProps
Moving Tooltip to wrap the IconButton

aachavarria committed:
Renaming BackButtonStack to PreviewBackButton
Renaming ForwardButtonStack to PreviewForwardButton
Renaming backStack to historyBackStack
Renaming forwardStack to historyForwardStack

aachavarria committed:
renaming useBackStack to useHistoryBackStack
renaming useForwardStack to useHistoryForwardStack

aachavarria committed:
Fixing issues related with the renaming

rart commented on ui/app/src/components/PreviewBackButton/PreviewBackButton.tsx:
```diff
@@ -0,0 +1,44 @@
+  };
+
+  return (
+    <Tooltip title={<FormattedMessage id="backButtonStack.back" defaultMessage="Back" />}>
```
```suggestion
    <Tooltip title={<FormattedMessage id="words.back" defaultMessage="Back" />}>
```

rart commented on ui/app/src/components/PreviewForwardButton/PreviewForwardButton.tsx:
```diff
@@ -0,0 +1,44 @@
+  };
+
+  return (
+    <Tooltip title={<FormattedMessage id="forwardButtonStack.forward" defaultMessage="Forward" />}>
```
```suggestion
    <Tooltip title={<FormattedMessage id="words.forward" defaultMessage="Forward" />}>
```

aachavarria committed:
Changing formattedMessage id
